libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,613 @@
+{
+    if(QSOLightupParams.QSOLightupOn)
+        load_heii_reion_hist(reion_hist_file);
+    N_ionized = 0;
+}
+
+static double last_zz;
+static double last_long_mfp_heating;
+
+/* Get the long mean free path heating. in erg/s/cm^3 */
+double
+get_long_mean_free_path_heating(double redshift)
+{
+    if(!QSOLightupParams.QSOLightupOn)
+        return 0;
+    if(redshift > QSOLightupParams.heIIIreion_start)
+        return 0;
+    if(redshift == last_zz)
+        return last_long_mfp_heating;
+    double atime = 1/(1+redshift);
+
+    /* Guard against the end of the table*/
+    if(atime > He_zz[Nreionhist-1])
+        return 0;
+
+    double long_mfp_heating = gsl_interp_eval(LMFP_intp, He_zz, LMFP, atime, NULL);
+
+    last_zz = redshift;
+    last_long_mfp_heating = long_mfp_heating;
+    return long_mfp_heating;
+}
+
+/* This function gets a random number from a Gaussian distribution using the Box-Muller transform.*/
+static double gaussian_rng(double mu, double sigma, const int64_t seed)
+{
+    double u1 = get_random_number(seed);
+    double u2 = get_random_number(seed + 1);
+    double z1 = sqrt(-2 * log(u1) ) * cos(2 * M_PI * u2);
+    return mu + sigma * z1;
+}
+
+/* Build a list of halos which are candidates for becoming a quasar.
+ * We use only halos with the right mass range.*/
+static int
+build_qso_candidate_list(int ** qso_cand, FOFGroups * fof)
+{
+    /* Run FOF to get a halo catalogue*/
+    /*Loop over all halos, building the candidate list.*/
+    int i, ncand=0;
+    *qso_cand = mymalloc("Quasar_candidates", sizeof(int) * (fof->Ngroups+1));
+    for(i = 0; i < fof->Ngroups; i++)
+    {
+        /* Check that it has the right mass*/
+        if(fof->Group[i].Mass < QSOLightupParams.qso_candidate_min_mass)
+            continue;
+        if(fof->Group[i].Mass > QSOLightupParams.qso_candidate_max_mass)
+            continue;
+        /*Add to the candidate list*/
+        (*qso_cand)[ncand] = i;
+        ncand++;
+    }
+    /*Poison value at the end for safety.*/
+    (*qso_cand)[ncand] = -1;
+    *qso_cand = myrealloc(*qso_cand, (ncand+1) * sizeof(int));
+    return ncand;
+}
+
+/* Choose a black hole particle at random to host a quasar from the list of black hole particles.
+ * This function chooses a single quasar from the total candidate list of quasars on *all* processors.
+ * We seed the random number generator off the number of existing quasars.
+ * This is done carefully so that we get the same sequence of quasars irrespective of how many processors we are using.
+ */
+static int
+choose_QSO_halo(int ncand, int64_t * ncand_tot, FOFGroups * fof, MPI_Comm Comm)
+{
+    int64_t ncand_total = 0, ncand_before = 0;
+    int NTask, i, ThisTask;
+    MPI_Comm_size(Comm, &NTask);
+    MPI_Comm_rank(Comm, &ThisTask);
+    int * candcounts = (int*) ta_malloc("qso_cand_counts", int, NTask);
+
+    /* Get how many candidates are on each processor*/
+    MPI_Allgather(&ncand, 1, MPI_INT, candcounts, 1, MPI_INT, Comm);
```

Comment:
Patterns like this for computing ncand_before and ncand_total has been around in many places. We probably want to make a function under MPIU for this. Could you add a 'TODO: add a function in MPIU' to remind us?